### PR TITLE
[rust] Refactor config

### PR DESCRIPTION
**Pull request by oriongonza**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Motivation and Context

This is the last PR in of in the series of refactors. This will allow to idiomatically implement newer flags such as `--cache-path`.
The goal of these changes is to introduce more idiomatic patterns into the codebase.

This includes 3 major changes:

- Every config string that could be `""` has been changed to an `Option`. This enforces explicit error handling, which is crucial when dealing with global mutable state as we do (I'm talking especially about you, driver version). This doesn't do much to help with the flow of state but at least it makes it explicit that the value **must exist** before it's called.
- The structs that get configuration from config file/env have been changed into functions. This simplifies the syntax and the semantics of it.
- Some clippy lints like string interpolation.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by diemol** [comment]
Seems the PR was created from an out of date branch. Can you please have a look before someone jumps into the review?

**Comment by oriongonza** [comment]
> Seems the PR was created from an out of date branch. Can you please have a look before someone jumps into the review?

I think I've merged the changes correctly.  What should I do in the future to avoid using an incorrect branch? I branched from trunk the day I did the changes.

**Comment by diemol** [comment]
> > Seems the PR was created from an out of date branch. Can you please have a look before someone jumps into the review?
> 
> I think I've merged the changes correctly. What should I do in the future to avoid using an incorrect branch? I branched from trunk the day I did the changes.

Usually take the most recent commit from `trunk` and start there. 

In addition, is there a way to reduce the size of the pull request? A pull request with 15 files changed takes longer to review, and sometimes one tends to delay the review because it just looks like too much to go through.

**Comment by oriongonza** [comment]
Hmm I think I could had made it smaller but this is one is tricky because it's a fairly big change that involves many files.

Some of those changes were the result of running `cargo fmt` after merging. I thought they belonged here but in hindsight it adds more to an already bloated PR.

File: rust/src/edge.rs [hunk 1]
```diff
@@ -180,23 +184,28 @@ impl SeleniumManager for EdgeManager {
...
-                    browser_version = self.get_major_version(latest_driver_version.as_str())?;
-                    self.log.debug(format!(
-                        "Latest {} major version is {}",
-                        &self.driver_name, browser_version
-                    ));
-                }
+                let browser_version = match browser_version {
+                    Some(version) => version.to_string(),
+                    _ => {
+                        let latest_stable_url = format!("{}{}", DRIVER_URL, LATEST_STABLE);
```
Comment (bonigarcia) [inline rust/src/edge.rs:190]: Since have changed this formatting, please do it consistently. For instance, this one should be:

```
format!("{DRIVER_URL}{LATEST_STABLE}");
```

Please review this in the rest of modules.

File: rust/src/config.rs [hunk 2]
```diff
@@ -145,59 +145,57 @@ impl ARCH {
...
+                .get(key)
+                .and_then(|x| x.as_str())
+                .map(|x| x.to_string());
+            if res.is_some() {
+                return res;
             }
+            env::var(get_env_name(key)).ok()
         }
-        self.1.to_owned()
+        Err(_) => None, // TODO log errors
```
Comment (bonigarcia) [inline rust/src/config.rs:165]: Please complete this `TODO`. The same for the other types.

File: rust/src/main.rs [hunk 3]
```diff
@@ -102,18 +102,16 @@ struct Cli {
 
 fn main() {
     let cli = Cli::parse();
-    let debug = cli.debug || BooleanKey("debug", false).get_value();
-    let trace = cli.trace || BooleanKey("trace", false).get_value();
-    let log = Logger::create(&cli.output, debug, trace);
+    let debug = cli.debug || boolean_key("debug").unwrap_or(false);
+    let trace = cli.trace || boolean_key("trace").unwrap_or(false);
```
Comment (bonigarcia) [inline rust/src/main.rs:106]: This new way to read envs or config file seem not to work. I tested manually by exporting the following env:

```
SE_TRACE=true
```

... and the executing selenium-manager as usual:

```
cargo run -- --browser chrome
```

When this feature works, the `TRACE` logs should be shown. But now only `INFO` (i.e., the default) are shown. 

The config file (`~/.cache/selenium/selenium-manager-config.toml`) should also work to provide extra configuration.

For more info, see #11810.

File: rust/tests/cli_tests.rs [hunk 4]
```diff
@@ -55,15 +54,18 @@ fn ok_test(
...
     };
-    println!("{}", output);
+    println!("{output}");
 
     assert!(output.contains(&driver_name));
     if !browser_version.is_empty() && output.contains("cache") {
         assert!(output.contains(&driver_version));
     }
-    assert!(!output.contains("Trying with latest driver version"));
+    assert!(
```
Comment (bonigarcia) [inline rust/tests/cli_tests.rs:65]: Please revert this change (that message is not necessary/accurate).

File: rust/src/edge.rs (hunk 1 above)
Comment (oriongonza) [reply rust/src/edge.rs:190]: That was done in commit e9c349911ab16c8f444f159650929c62a52ddb98.

File: rust/tests/cli_tests.rs (hunk 4 above)
Comment (oriongonza) [reply rust/tests/cli_tests.rs:65]: Hm but it is strange. In my machine that test failed until I installed firefox.

**Comment by oriongonza** [comment]
I will not be able to work on this until at least next week Tuesday. I just hope the merge conflicts don't get too out of hand...

I plan to split this PR into two, I'll close this once they are up.

https://github.com/SeleniumHQ/selenium/pull/11921#discussion_r1178290854 This bothers me, it's entirely possible that I messed up but it's an odd thing.
